### ISSUE TITLE
Questions on 09_up_and_running_with_tensorflow.ipynb

### ISSUE THREAD

**Issue by tongjiyiming** [ageron/handson-ml#115]
I tried my own version of mini-batch part. Basically, I loop through each index to get the batch. I can see that your solution use a random selected index of batch size. The result of trained value of theta is similar. But, I wonder why you use that way, and which way is more suitable in terms of accuracy, and in terms of practice?

Thank you!

here is my code for the function fetch_batch():
```python
def fetch_batch(epoch, batch_index, batch_size):
    start_index = batch_index*batch_size
    end_index = start_index + batch_size
    if end_index > len(scaled_housing_data):
        end_index = len(scaled_housing_data)
        
    X_batch = np.c_[ np.ones((end_index-start_index, 1)), scaled_housing_data[start_index:end_index] ]
    y_batch = housing.target.reshape(-1, 1)[start_index:end_index]
    return X_batch, y_batch
```

**Comment by ageron** [comment]
Hi @tongjiyiming ,
Interesting question.  In general, it's better to shuffle the training data at each epoch, to ensure that there is no correlation between instances within a training batch.  If the data is already shuffled before you start training, then it may not make a huge difference to shuffle the data again at each epoch, but it's still good practice.
For example, let's try training a classifier on MNIST, either on a sorted dataset (first all 0s, then all 1s, and so on), or on a shuffled dataset, and let's look at the resulting accuracy in both cases:

```python
from __future__ import print_function
from sklearn.datasets import fetch_mldata
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
mnist = fetch_mldata('MNIST original')
X_train, X_test = mnist.data[:60000], mnist.data[60000:]
y_train, y_test = mnist.target[:60000], mnist.target[60000:]
for shuffle in (False, True):
    print("Training with shuffle =", shuffle)
    sgd = SGDClassifier(shuffle=shuffle, max_iter=10)
    sgd.fit(X_train, y_train)
    y_pred = sgd.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print("Accuracy =", accuracy)
```

The output is:
```
Training with shuffle = False
Accuracy = 0.1155
Training with shuffle = True
Accuracy = 0.8841
```

The difference is huge! The model trained on sorted data is not much better than a random model, while the one trained on shuffled data reaches 88.4% accuracy.
